trio/_tests/check_type_completeness.py

```diff
@@ -0,0 +1,160 @@
+
+    # check if py.typed file was missing
+    if (
+        current_result["generalDiagnostics"]
+        and current_result["generalDiagnostics"][0]["message"]
+        == "No py.typed file found"
+    ):
+        print("creating py.typed")
+        py_typed_file = (
+            Path(current_result["typeCompleteness"]["packageRootDirectory"])
+            / "py.typed"
+        )
+        py_typed_file.write_text("")
+
+        res = run_pyright()
+        current_result = json.loads(res.stdout)
+
+    if res.stderr:
+        print(res.stderr)
+
+    last_result = json.loads(RESULT_FILE.read_text())
+
+    for key in "errorCount", "warningCount", "informationCount":
+        check_zero(key, current_result["summary"])
+
+    for key, invert in (
+        ("missingFunctionDocStringCount", False),
+        ("missingClassDocStringCount", False),
+        ("missingDefaultParamCount", False),
+        ("completenessScore", True),
+    ):
+        check_less_than(
+            key,
+            current_result["typeCompleteness"],
+            last_result["typeCompleteness"],
+            invert=invert,
+        )
+
+    for key, invert in (
+        ("withUnknownType", False),
+        ("withAmbiguousType", False),
+        ("withKnownType", True),
+    ):
+        check_less_than(
+            key,
+            current_result["typeCompleteness"]["exportedSymbolCounts"],
+            last_result["typeCompleteness"]["exportedSymbolCounts"],
+            invert=invert,
+        )
+
+    assert (
+        res.returncode != 0
+    ), "Fully type complete! Delete this script and instead directly run `pyright --verifytypes=trio` in CI and checking exit code."
+
+    if args.overwrite_file:
+        print("Overwriting file")
+
+        # don't care about differences in time taken
+        del current_result["time"]
+        del current_result["summary"]["timeInSec"]
+
+        # don't fail on version diff so pyright updates can be automerged
+        del current_result["version"]
+
+        for key in (
+            # don't save path (because that varies between machines)
+            "moduleRootDirectory",
+            "packageRootDirectory",
+            "pyTypedPath",
+        ):
+            del current_result["typeCompleteness"][key]
+
+        new_symbols = []
+        # remove path & line/col references in errors
+        for symbol in current_result["typeCompleteness"]["symbols"]:
+            # filter out symbols with no errors
+            if not symbol["diagnostics"]:
+                continue
+            for diag in symbol["diagnostics"]:
+                del diag["file"]
+                diag.pop("range", None)
+            new_symbols.append(symbol)
+
+        current_result["typeCompleteness"]["symbols"] = new_symbols
+
+        with open(RESULT_FILE, "w") as file:
+            json.dump(current_result, file, sort_keys=True, indent=2)
+            # add newline at end of file so it's easier to manually modify
+            file.write("\n")
```

Comment:
OK I reread things and I kinda don't get why we keep `symbols` at all. I don't see them being used elsewhere, am I missing something? They're a *gigantic* list and probably most of the reason the file is as big as it is.

Reply:
We initially removed it, but I realized when I was trying out stuff that without it you won't get any feedback if/when the check fails. The numbers will change, but unless you `checkout` previous revisions, manually run `pyright --verifytypes`, pipe into a file, and manually diff against a manual `--verifytypes` run on your PR - you're left guessing what symbol you broke where. And if your local environment differs from CI, then you're absolutely toast. (I have recurring problems with CI not giving same results as running stuff in my virtualenv and/or tox envs)

I could shorten the file further, removing more info that's probably redundant, not pretty-print it - or like dump `symbols` into a compressed pickle that the script reads to say what symbols have changed, but:
1. the json file is not included when packaging
2. the json file will likely drop down in size a lot early on as I do a first pass and introduce a lot of trivial typing on functions.

If you still think it's a problem, I'll make a pass at sizing it down. Or I can toss symbols out for now, run the script modified locally, and after (2) when the file is smaller add it to the repo.

Other esoteric options include commiting the file to a different branch, git-annex, have it in a submodule...

Reply:
Hmm, my concern is that we're making the repository larger for anyone who wants a (full) clone. Do you have measurements of the .git folder before and after this? (You may want to rebase first to get rid of the old much bigger json file)

Reply:
Looks like going from something like 7.9M -> 8.1M
But it's a good point, I hadn't realized that big files will stick around forever in the history. I'll test pruning it down to being just a list of symbols with no extra info